lib/Conversion/Utils.cpp

```diff
@@ -12,6 +16,237 @@ using ::mlir::func::CallOp;
+  // where SourceType would be type-converted to tensor<...> to
+  // a concatenation of the converted operands (with appropriate reshape)
+  LogicalResult matchAndRewrite(
+      tensor::FromElementsOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    // Expand each of the (converted) operands:
+    SmallVector<Value> newOperands;
+    for (auto o : adaptor.getElements()) {
+      // extend tensor<...xT> to tensor<1x...xT>
+      if (auto tensorType = o.getType().dyn_cast<RankedTensorType>()) {
+        auto shape = tensorType.getShape();
+        SmallVector<int64_t> newShape(1, 1);
+        newShape.append(shape.begin(), shape.end());
+
+        // Create a dense constant for targetShape
+        auto shapeOp = rewriter.create<arith::ConstantOp>(
+            op.getLoc(),
+            RankedTensorType::get(newShape.size(), rewriter.getIndexType()),
+            rewriter.getIndexTensorAttr(newShape));
+
+        auto reshapeOp = rewriter.create<tensor::ReshapeOp>(
+            op.getLoc(),
+            RankedTensorType::get(newShape, tensorType.getElementType()), o,
+            shapeOp);
+        newOperands.push_back(reshapeOp);
+      } else {
+        newOperands.push_back(o);
+      }
+    }
+    // Create the final tensor.concat operation
+    rewriter.replaceOpWithNewOp<tensor::ConcatOp>(op, 0, newOperands);
+
+    return success();
+  }
+};
+
+void addTensorOfTensorConversionPatterns(TypeConverter &typeConverter,
+                                         RewritePatternSet &patterns,
+                                         ConversionTarget &target) {
+  typeConverter.addConversion([&](TensorType type) -> Type {
+    if (!typeConverter.isLegal(type.getElementType())) {
+      typeConverter.convertType(type.getElementType()).dump();
+      if (auto convertedType =
+              typeConverter.convertType(type.getElementType())) {
+        if (auto castConvertedType =
+                convertedType.dyn_cast<RankedTensorType>()) {
+          //  Create the combined shape
+          auto polyShape = castConvertedType.getShape();
+          auto tensorShape = type.getShape();
+          SmallVector<int64_t, 4> combinedShape(tensorShape.begin(),
+                                                tensorShape.end());
+          combinedShape.append(polyShape.begin(), polyShape.end());
+          auto combinedType = RankedTensorType::get(
+              combinedShape, castConvertedType.getElementType());
+          return combinedType;
+        }
+      }
+    }
+    return type;
+  });
+
+  target.addDynamicallyLegalDialect<tensor::TensorDialect>([&](Operation *op) {
+    return typeConverter.isLegal(op->getOperandTypes()) &&
+           typeConverter.isLegal(op->getResultTypes());
+  });
+
+  target.addDynamicallyLegalDialect<affine::AffineDialect>([&](Operation *op) {
+    return typeConverter.isLegal(op->getOperandTypes()) &&
+           typeConverter.isLegal(op->getResultTypes());
+  });
```

Comment:
In the meantime, I noticed & fixed a minor bug that meant ops like `affine.yield` weren't getting handled properly (popped up while testing the bgv->poly->standard chain) because the legality check previously only considered the OperandTypes but not the ResultTypes.

Reply:
It looks like the original PR before this fix was merged asyncronously to your fix as 348c2252cf44e0378a64545dbeec682a51d83dd1 (which appears to be why this PR was not automatically marked as merged). Could you cherry pick that change to a new PR and close this one?